File: projects/subscription-manager/frontend/src/types/subscription.ts
export type BillingCycle = 'weekly' | 'monthly' | 'yearly' | 'custom';

export type Category =
  | 'entertainment'
  | 'productivity'
  | 'lifestyle'
  | 'utilities'
  | 'news'
  | 'education'
  | 'finance'
  | 'health'
  | 'other';

export type Currency = 'USD' | 'EUR' | 'GBP' | 'CAD' | 'AUD' | 'INR';

export interface Subscription {
  id: string;
  name: string;
  description?: string;
  price: number;
  currency: Currency;
  billingCycle: BillingCycle;
  customDays?: number;
  nextBillingDate: string;
  category: Category;
  icon?: string;
  color?: string;
  notes?: string;
  isActive: boolean;
  reminderDays: number;
  createdAt: string;
  updatedAt: string;
}

export interface SubscriptionInput {
  name: string;
  description?: string;
  price: number;
  currency?: Currency;
  billingCycle: BillingCycle;
  customDays?: number;
  nextBillingDate: string;
  category?: Category;
  icon?: string;
  color?: string;
  notes?: string;
  reminderDays?: number;
}

export interface SubscriptionTemplate {
  name: string;
  icon: string;
  color: string;
  category: Category;
  defaultPrice: number;
  billingCycle: BillingCycle;
}

export const CATEGORIES: { value: Category; label: string; icon: string }[] = [
  { value: 'entertainment', label: 'Entertainment', icon: '🎬' },
  { value: 'productivity', label: 'Productivity', icon: '⚡' },
  { value: 'lifestyle', label: 'Lifestyle', icon: '🌟' },
  { value: 'utilities', label: 'Utilities', icon: '🔧' },
  { value: 'news', label: 'News', icon: '📰' },
  { value: 'education', label: 'Education', icon: '📚' },
  { value: 'finance', label: 'Finance', icon: '💰' },
  { value: 'health', label: 'Health', icon: '❤️' },
  { value: 'other', label: 'Other', icon: '📦' },
];

export const BILLING_CYCLES: { value: BillingCycle; label: string }[] = [
  { value: 'weekly', label: 'Weekly' },
  { value: 'monthly', label: 'Monthly' },
  { value: 'yearly', label: 'Yearly' },
  { value: 'custom', label: 'Custom' },
];

export function getMonthlyEquivalent(sub: Subscription): number {
  switch (sub.billingCycle) {
    case 'weekly':
      return (sub.price * 52) / 12;
    case 'monthly':
      return sub.price;
    case 'yearly':
      return sub.price / 12;
    case 'custom':
      if (sub.customDays && sub.customDays > 0) {
        return (sub.price * 30) / sub.customDays;
      }
      return sub.price;
    default:
      return sub.price;
  }
}

export function getYearlyEquivalent(sub: Subscription): number {
  switch (sub.billingCycle) {
    case 'weekly':
      return sub.price * 52;
    case 'monthly':
      return sub.price * 12;
    case 'yearly':
      return sub.price;
    case 'custom':
      if (sub.customDays && sub.customDays > 0) {
        return (sub.price * 365) / sub.customDays;
      }
      return sub.price * 12;
    default:
      return sub.price * 12;
  }
}

export function getDaysUntilBilling(nextBillingDate: string): number {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  const billing = new Date(nextBillingDate);
  billing.setHours(0, 0, 0, 0);
  return Math.ceil((billing.getTime() - today.getTime()) / (1000 * 60 * 60 * 24));
}

export function isDueSoon(nextBillingDate: string, days: number = 7): boolean {
  const daysUntil = getDaysUntilBilling(nextBillingDate);
  return daysUntil >= 0 && daysUntil <= days;
}

export function formatBillingCycle(cycle: BillingCycle): string {
  switch (cycle) {
    case 'weekly':
      return '/wk';
    case 'monthly':
      return '/mo';
    case 'yearly':
      return '/yr';
    default:
      return '';
  }
}

export function formatDate(dateString: string): string {
  const date = new Date(dateString);
  return date.toLocaleDateString('en-US', {
    month: 'short',
    day: 'numeric',
  });
}

export function formatFullDate(dateString: string): string {
  const date = new Date(dateString);
  return date.toLocaleDateString('en-US', {
    weekday: 'long',
    month: 'long',
    day: 'numeric',
    year: 'numeric',
  });
}

File: projects/subscription-manager/backend/internal/model/subscription.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type BillingCycle string

const (
	BillingWeekly  BillingCycle = "weekly"
	BillingMonthly BillingCycle = "monthly"
	BillingYearly  BillingCycle = "yearly"
	BillingCustom  BillingCycle = "custom"
)

type Category string

const (
	CategoryEntertainment Category = "entertainment"
	CategoryProductivity  Category = "productivity"
	CategoryLifestyle     Category = "lifestyle"
	CategoryUtilities     Category = "utilities"
	CategoryNews          Category = "news"
	CategoryEducation     Category = "education"
	CategoryFinance       Category = "finance"
	CategoryHealth        Category = "health"
	CategoryOther         Category = "other"
)

type Subscription struct {
	ID              uuid.UUID    `json:"id"`
	UserID          uuid.UUID    `json:"userId"`
	Name            string       `json:"name"`
	Description     *string      `json:"description,omitempty"`
	Price           float64      `json:"price"`
	Currency        string       `json:"currency"`
	BillingCycle    BillingCycle `json:"billingCycle"`
	CustomDays      *int         `json:"customDays,omitempty"`
	NextBillingDate time.Time    `json:"nextBillingDate"`
	Category        Category     `json:"category"`
	Icon            *string      `json:"icon,omitempty"`
	Color           *string      `json:"color,omitempty"`
	Notes           *string      `json:"notes,omitempty"`
	IsActive        bool         `json:"isActive"`
	ReminderDays    int          `json:"reminderDays"`
	CreatedAt       time.Time    `json:"createdAt"`
	UpdatedAt       time.Time    `json:"updatedAt"`
}

type SubscriptionInput struct {
	Name            string       `json:"name"`
	Description     *string      `json:"description,omitempty"`
	Price           float64      `json:"price"`
	Currency        *string      `json:"currency,omitempty"`
	BillingCycle    BillingCycle `json:"billingCycle"`
	CustomDays      *int         `json:"customDays,omitempty"`
	NextBillingDate string       `json:"nextBillingDate"`
	Category        *Category    `json:"category,omitempty"`
	Icon            *string      `json:"icon,omitempty"`
	Color           *string      `json:"color,omitempty"`
	Notes           *string      `json:"notes,omitempty"`
	ReminderDays    *int         `json:"reminderDays,omitempty"`
}

func (s *Subscription) MonthlyEquivalent() float64 {
	switch s.BillingCycle {
	case BillingWeekly:
		return s.Price * 52 / 12
	case BillingMonthly:
		return s.Price
	case BillingYearly:
		return s.Price / 12
	case BillingCustom:
		if s.CustomDays != nil && *s.CustomDays > 0 {
			return s.Price * 30 / float64(*s.CustomDays)
		}
		return s.Price
	default:
		return s.Price
	}
}

func (s *Subscription) YearlyEquivalent() float64 {
	switch s.BillingCycle {
	case BillingWeekly:
		return s.Price * 52
	case BillingMonthly:
		return s.Price * 12
	case BillingYearly:
		return s.Price
	case BillingCustom:
		if s.CustomDays != nil && *s.CustomDays > 0 {
			return s.Price * 365 / float64(*s.CustomDays)
		}
		return s.Price * 12
	default:
		return s.Price * 12
	}
}

func (s *Subscription) DaysUntilBilling() int {
	now := time.Now().Truncate(24 * time.Hour)
	billing := s.NextBillingDate.Truncate(24 * time.Hour)
	return int(billing.Sub(now).Hours() / 24)
}

func (s *Subscription) IsDueSoon(days int) bool {
	return s.DaysUntilBilling() <= days && s.DaysUntilBilling() >= 0
}
